File: packages/ruby/lib/groundcontrolsh.rb
require 'uri'
require 'net/http'
require 'openssl'
require 'json'
require 'cgi'

class GroundControl
  def initialize(project_id:, api_key:, **options)
    @project_id = project_id
    @api_key = api_key
    @base_url = options[:base_url] || 'https://api.groundcontrol.sh'
    @ttl = options[:cache] || nil
    @cache = {}

    @_actor_overrides = {}
    @_flag_overrides = {}
    @full_override = nil
  end

  def feature_flag_enabled?(flag_name, options = {})
    actors = options[:actors] || []

    actor_overrides = @_actor_overrides[flag_name]
    actors.each do |actor_id|
      override = actor_overrides[actor_id]
      return override if override != nil
    end if actor_overrides != nil && actors != nil

    flag_override = @_flag_overrides[flag_name]
    return flag_override if flag_override != nil
    return @_full_override if @_full_override != nil

    query = actors.map { |actor| "actorIds=#{CGI.escape(actor)}" }
    url = URI("#{@base_url}/projects/#{@project_id}/flags/#{flag_name}/check?#{query.join('&')}")
    cached = @cache[url]
    if cached
      if cached[:expires] < Time.now.to_i
        @cache.delete(url)
      else
        return cached[:enabled]
      end
    end

    http = Net::HTTP.new(url.host, url.port)
    http.use_ssl = true

    request = Net::HTTP::Get.new(url)
    request["Authorization"] = "Bearer #{@api_key}"
    response = http.request(request)

    return nil if response.code != '200'

    body = JSON.parse(response.body)
    enabled = body["enabled"]

    if @ttl
      @cache[url] = { expires: Time.now.to_i + @ttl, enabled: enabled }
    end

    return enabled
  end

  def disable_feature_flag(flag_name, options = nil)
    set_feature_flag_enabled(false, flag_name, options)
  end

  def disable_all_feature_flags
    @_full_override = false
  end

  def enable_feature_flag(flag_name, options = nil)
    set_feature_flag_enabled(true, flag_name, options)
  end

  def enable_all_feature_flags()
    @_full_override = true
  end

  def reset
    @_actor_overrides.clear
    @_flag_overrides.clear
    @_full_override = nil
  end

  private

  def set_feature_flag_enabled(enabled, flag_name, options)
    options = {} if options == nil
    actors_ops = options[:actors]
    if actors_ops == nil
      @_flag_overrides[flag_name] = enabled
    else
      actors = @_actor_overrides[flag_name] || {}
      actors_ops.each do |actor_id|
        actors[actor_id] = enabled
      end
      @_actor_overrides[flag_name] = actors
    end
  end
end

File: packages/python/src/groundcontrolsh/groundcontrolsh.py
import requests
import time
from urllib.parse import quote
import typing as t

class GroundControl:

    def __init__(self, project_id: str, api_key: str, **options: t.Any) -> None:
        self.project_id: str = project_id
        self.api_key: str = api_key
        self.base_url: str = options.get('base_url', 'https://api.groundcontrol.sh')
        self.ttl: int | None = options.get('cache')
        self.cache: t.Dict[str, t.Dict[str, t.Any]] = {}

        self.actor_overrides: t.Dict[str, t.Dict[str, bool]] = {}
        self.flag_overrides: t.Dict[str, bool] = {}
        self.full_override: bool | None = None

    def is_feature_flag_enabled(self, flag_name: str, options: t.Dict[str, t.Any] | None = None) -> bool:
        options = options or {}
        actors = options.get('actors', [])

        actor_overrides = self.actor_overrides.get(flag_name, {})
        for actor_id in actors:
            override = actor_overrides.get(actor_id)
            if override is not None:
                return override

        flag_override = self.flag_overrides.get(flag_name)
        if flag_override is not None:
            return flag_override

        if self.full_override is not None:
            return self.full_override

        query = "&".join(f"actorIds={quote(actor)}" for actor in actors)
        url = f"{self.base_url}/projects/{self.project_id}/flags/{flag_name}/check?{query}"
        cached = self.cache.get(url)

        if cached:
            if cached['expires'] < int(time.time()):
                self.cache.pop(url)
            else:
                return cached['enabled']

        headers = {"Authorization": f"Bearer {self.api_key}"}
        response = requests.get(url, headers=headers)

        if response.status_code != 200:
            # TODO: use on_error callback
            return False

        body = response.json()
        enabled = body["enabled"]

        if self.ttl:
            self.cache[url] = {
                "expires": int(time.time()) + self.ttl,
                "enabled": enabled
            }

        return enabled

    def disable_feature_flag(self, flag_name: str, options: t.Dict[str, t.Any] | None = None) -> None:
        self.set_feature_flag_enabled(False, flag_name, options)

    def disable_all_feature_flags(self) -> None:
        self.full_override = False

    def enable_feature_flag(self, flag_name: str, options: t.Dict[str, t.Any] | None = None) -> None:
        self.set_feature_flag_enabled(True, flag_name, options)

    def enable_all_feature_flags(self) -> None:
        self.full_override = True

    def reset(self) -> None:
        self.actor_overrides.clear()
        self.flag_overrides.clear()
        self.full_override = None

    def set_feature_flag_enabled(self, enabled, flag_name: str, options: t.Dict[str, t.Any] | None = None) -> None:
        options = options or {}
        actors_ops = options.get('actors')

        if actors_ops is None:
            self.flag_overrides[flag_name] = enabled
        else:
            actors = self.actor_overrides.get(flag_name, {})
            for actor_id in actors_ops:
                actors[actor_id] = enabled
            self.actor_overrides[flag_name] = actors
